### PR TITLE
Add progress reporting to the TS client

### LINKED ISSUE #133: Support progress in django-s3-file-field-client
I am using this to upload some large datasets (many GB at a time), so having real upload progress will be important. As a user I would imagine this would be a callback I could optionally pass to `uploadFile`.

### PR DESCRIPTION
**Pull request by brianhelba**
Fixes #133.

### CODE REVIEW

**Comment by brianhelba** [comment]
@zachmullen @dchiquit This is an alternative to #134, which should be safe for use with parallel uploads.

File: django-s3-file-field-client/src/client.ts [hunk 1]
```diff
@@ -25,11 +31,51 @@ export interface UploadResult {
   state: 'aborted' | 'successful' | 'error';
 }
 
-export default class S3FFClient {
-  protected baseUrl: string;
+class Progressable {
+  private overheadPercentage = 0.0;
+  private mainPercentageSegments: number[] = [];
+  // This value was picked arbitrarily
+  private readonly overheadProportion = 0.20;
```
Comment (zachmullen) [inline django-s3-file-field-client/src/client.ts:38]: I'm not fond of this statically fixed overhead proportion. If we insist on not using indeterminate progress, let's at least base the proportion intelligently on the file's size.

File: django-s3-file-field-client/src/client.ts [hunk 2]
```diff
@@ -1,5 +1,11 @@
 import axios from 'axios';
 
+export interface Progress {
+  readonly percentage: number;
```
Comment (zachmullen) [inline django-s3-file-field-client/src/client.ts:4]: This value ranges `[0, 1]`, not `[0, 100]`, right? In that case, percentage is the wrong name. Maybe `fraction`, `proportion`, `ratio`... some other synonym?

File: django-s3-file-field-client/src/client.ts [hunk 3]
```diff
@@ -25,11 +31,51 @@ export interface UploadResult {
...
+  protected updateMainProgress(segment: number, current: number, total: number) {
+    // This weights each segment equally, which is not as accurate, but easier to initialize
+    this.mainPercentageSegments[segment] = current / total;
+    this.updateProgress();
+  }
+}
+
+export default class S3FFClient extends Progressable{
+  constructor(
+    protected readonly baseUrl: string,
```
Comment (brianhelba) [inline django-s3-file-field-client/src/client.ts:75]: Note this TypeScript feature for declaring and initializing member variables in one line. I've updated `baseUrl` to do that.

File: django-s3-file-field-client/src/client.ts [hunk 4]
```diff
@@ -25,11 +31,51 @@ export interface UploadResult {
...
+    }
+  }
+
+  protected updateOverheadProgress(percentage: number) {
+    this.overheadPercentage = percentage;
+    this.updateProgress();
+  }
+
+  protected initializeMainProgress(segments: number) {
+    this.mainPercentageSegments = new Array(segments).fill(0);
```
Comment (zachmullen) [inline django-s3-file-field-client/src/client.ts:63]: I wrote up (but haven't yet tested) an alternative implementation that uses less memory. It leverages closures to track the progress incrementally:

```javascript
protected async uploadPart(chunk: ArrayBuffer, part: PartInfo, incrementProgress: Function): Promise<UploadedPart> {
    let lastLoaded = 0;
    const response = await axios.put(part.upload_url, chunk, {
      onUploadProgress: (e) => {
        const increment: number = e.loaded - lastLoaded;
        lastLoaded = e.loaded;
        incrementProgress(increment);
      },
    });
    const etag = response.headers.etag;
    return {
      part_number: part.part_number,
      size: part.size,
      etag,
    };
  }
```

...

```javascript
protected async uploadParts(file: File, parts: PartInfo[]): Promise<UploadedPart[]> {
    const buffer = await file.arrayBuffer();
    // indices track where in the buffer each part begins
    let index = 0;
    const indices: number[] = [];
    for (const part of parts) {
      indices.push(index);
      index += part.size;
    }

    let totalLoaded = 0;
    const incrementProgress = (increment: number) => {
      totalLoaded += increment;
      this.progressCallback({
        loaded: totalLoaded,
        total: file.size,
        lengthComputable: true,
      });
    };

    // upload each part of the buffer in parallel using the calculated indices
    return await Promise.all(parts.map(async (part, i) => {
      const chunk = buffer.slice(indices[i], indices[i] + part.size);
      return await this.uploadPart(chunk, part, incrementProgress);
    }));
  }
```

What do you think?

**Review by zachmullen — CHANGES_REQUESTED** [review]

**Comment by brianhelba** [comment]
Superseded by #164.

**Pull request closed without merging**